sklearn/linear_model/coordinate_descent.py

```diff
@@ -2033,3 +2035,222 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
             max_iter=max_iter, tol=tol, copy_X=copy_X,
             cv=cv, verbose=verbose, n_jobs=n_jobs, random_state=random_state,
             selection=selection)
+
+
+class AdaptiveLasso(Lasso):
+    """The multi-step adaptive Lasso iteratively solves Lasso estimates with
+    penalty weights applied to the regularization of the coefficients.
+
+    The optimization objective for the AdaptiveLasso is::
+
+        (1 / (2 * n_samples)) * ||y - X Beta||^2_2
+        + alpha * p(|Beta|_1 + eps)
```

Comment:
you should parametrize it with:

alpha * (|Beta|_1 + eps)^p

and then the solver to use p=0.5 by default. It's easier to think of the exponent of the reg than its derivative.